src/VecSim/algorithms/hnsw/hnsw.h

```diff
@@ -988,7 +988,18 @@ void HNSWIndex<DataType, DistType>::repairConnectionsForDeletion(
         auto &not_chosen_candidates = candidate_ids; // rename and reuse the vector
         getNeighborsByHeuristic2(candidates, Mcurmax, not_chosen_candidates);
 
-        neighbor_level.setLinks(candidates);
+        size_t new_num_links = 0;
+        for (auto chosen_cand : candidates) {
+            // We will add the selected element as a new neighbor ONLY IF:
+            // 1. The selected element was already a connection before - hence, no update performed.
+            // 2. Both the neighbour which we are updating and the selected element to connect are
+            // not deleted, hence we do not create new connections for marked deleted elements.
+            if (neighbour_orig_neighbours_set[chosen_cand.second] ||
+                (!isMarkedDeleted(neighbour_id) && !isMarkedDeleted(chosen_cand.second))) {
+                neighbor_level.setLinkAtPos(new_num_links++, chosen_cand.second);
+            }
+        }
+        neighbor_level.setNumLinks(new_num_links);
```

Comment:
aren't we locked here when performing this change? Do we have to check the deletion mark on every iteration?